File: frontend/src/api/sync.ts
import { ApiError, apiRequest } from '@/api/client'
import { isSyncEntityType, type SyncEntityType } from '@/store/syncMeta'

/**
 * The synchronisation endpoints.
 *
 * The server stores opaque envelopes and arbitrates on plaintext metadata
 * alone — a revision, a client timestamp, a tombstone flag. It never sees a
 * name, a note or a tag.
 */

/** Server-side ceilings. Exceeding one is a 422, so the engine batches to them. */
export const MAX_RECORDS_PER_PUSH = 500
export const PULL_PAGE_SIZE = 200

export interface PushRecord {
  entityType: SyncEntityType
  entityId: string
  /** The revision this device last saw. Null on a record it believes is new. */
  baseRevision: number | null
  clientUpdatedAt: string
  deleted: boolean
  ciphertext?: string
  nonce?: string
}

export interface RecordEnvelope {
  entityType: SyncEntityType
  entityId: string
  revision: number
  clientUpdatedAt: string
  deleted: boolean
  ciphertext: string | null
  nonce: string | null
}

export interface AppliedRecord {
  entityType: SyncEntityType
  entityId: string
  revision: number
}

export interface PullResponse {
  records: RecordEnvelope[]
  nextCursor: number
  hasMore: boolean
}

export interface PushResponse {
  applied: AppliedRecord[]
  conflicts: RecordEnvelope[]
}

export interface SyncStatus {
  serverRevision: number
  recordCount: number
}

/**
 * Checked rather than cast, because these responses drive destructive local
 * writes: an envelope read as a tombstone deletes a record, and a cursor read
 * as a number decides which records this device will never ask for again. A
 * malformed body has to fail loudly, before any of that.
 */
function isEnvelope(value: unknown): value is RecordEnvelope {
  if (value === null || typeof value !== 'object') return false
  const { entityType, entityId, revision, clientUpdatedAt, deleted } = value as Record<
    string,
    unknown
  >
  return (
    isSyncEntityType(entityType) &&
    typeof entityId === 'string' &&
    entityId !== '' &&
    typeof revision === 'number' &&
    typeof clientUpdatedAt === 'string' &&
    typeof deleted === 'boolean'
  )
}

function isApplied(value: unknown): value is AppliedRecord {
  if (value === null || typeof value !== 'object') return false
  const { entityType, entityId, revision } = value as Record<string, unknown>
  return (
    isSyncEntityType(entityType) && typeof entityId === 'string' && typeof revision === 'number'
  )
}

/**
 * Well-formed, but of a kind this version of the app has no list for — what a
 * newer client on the same account would push.
 *
 * Dropped here rather than refused: a record nothing on this device can hold
 * is not a malformed response, and treating it as one would strand every
 * record on the same page, permanently, behind a kind that will keep coming
 * back. The engine then treats what is left exactly as it treats an envelope
 * it cannot decrypt — the carnet is not touched, and nothing is claimed about
 * a version this device does not hold.
 */
function isForeignKind(value: unknown): boolean {
  if (value === null || typeof value !== 'object') return false
  const { entityType } = value as Record<string, unknown>
  return typeof entityType === 'string' && !isSyncEntityType(entityType)
}

/** Drops the kinds this version cannot hold, then insists on the rest. */
function checked<T>(values: unknown[], is: (value: unknown) => value is T): T[] {
  const kept = values.filter((value) => !isForeignKind(value))
  const valid = kept.filter(is)
  if (valid.length !== kept.length) malformed()
  return valid
}

function malformed(): never {
  throw new ApiError(502, 'Réponse de synchronisation inattendue.')
}

export async function pullChanges(since: number, limit = PULL_PAGE_SIZE): Promise<PullResponse> {
  const body = await apiRequest<unknown>(`/sync/changes?since=${since}&limit=${limit}`)
  if (body === null || typeof body !== 'object') malformed()
  const { records, nextCursor, hasMore } = body as Record<string, unknown>
  if (!Array.isArray(records)) malformed()
  const page = checked(records, isEnvelope)
  if (typeof nextCursor !== 'number' || typeof hasMore !== 'boolean') malformed()
  // The cursor still moves past the dropped ones: it is the server's, and
  // holding it back would re-read the same page for as long as they exist.
  return { records: page, nextCursor, hasMore }
}

export async function pushChanges(records: PushRecord[]): Promise<PushResponse> {
  const body = await apiRequest<unknown>('/sync/changes', { method: 'POST', body: { records } })
  if (body === null || typeof body !== 'object') malformed()
  const { applied, conflicts } = body as Record<string, unknown>
  if (!Array.isArray(applied) || !Array.isArray(conflicts)) malformed()
  return { applied: checked(applied, isApplied), conflicts: checked(conflicts, isEnvelope) }
}

export function fetchSyncStatus(): Promise<SyncStatus> {
  return apiRequest<SyncStatus>('/sync/status')
}

File: backend/app/schemas/sync.py
from datetime import datetime
from typing import Annotated, Literal, Self

from pydantic import AwareDatetime, Base64Bytes, Field, model_validator

from app.schemas.base import Base64Out, CarnetSchema

# The seven kinds of record a carnet holds. A Literal rather than a free string:
# the server cannot read the envelopes, so this list is the only structural
# check it can make on what it is asked to store.
EntityType = Literal[
    "etablissement",
    "classe",
    "eleve",
    "tagCategory",
    "tag",
    "event",
    "preference",
]

MAX_CIPHERTEXT_BYTES = 64 * 1024
MAX_RECORDS_PER_PUSH = 500
MAX_PULL_LIMIT = 500
DEFAULT_PULL_LIMIT = 200

EntityId = Annotated[str, Field(min_length=1, max_length=64)]
Ciphertext = Annotated[Base64Bytes, Field(min_length=1, max_length=MAX_CIPHERTEXT_BYTES)]
Nonce = Annotated[Base64Bytes, Field(min_length=12, max_length=16)]


class PushRecord(CarnetSchema):
    entity_type: EntityType
    entity_id: EntityId
    # The revision this device last saw. None on a record it believes is new.
    base_revision: int | None = None
    client_updated_at: AwareDatetime
    deleted: bool = False
    ciphertext: Ciphertext | None = None
    nonce: Nonce | None = None

    @model_validator(mode="after")
    def envelope_matches_the_tombstone_flag(self) -> Self:
        """A record is either an envelope or a tombstone, never both or neither.

        Without this, a client bug could store a live record with no contents,
        which no device could ever decrypt and none could tell from a deletion.
        """
        if self.deleted:
            if self.ciphertext is not None or self.nonce is not None:
                raise ValueError("Un enregistrement supprimé ne porte ni ciphertext ni nonce.")
            return self
        if self.ciphertext is None or self.nonce is None:
            raise ValueError("Un enregistrement vivant doit porter un ciphertext et un nonce.")
        return self


class PushRequest(CarnetSchema):
    records: Annotated[list[PushRecord], Field(min_length=1, max_length=MAX_RECORDS_PER_PUSH)]


class RecordEnvelope(CarnetSchema):
    """What the server hands back — the same opaque blob it was given."""

    entity_type: str
    entity_id: str
    revision: int
    client_updated_at: datetime
    deleted: bool
    ciphertext: Base64Out | None = None
    nonce: Base64Out | None = None


class AppliedRecord(CarnetSchema):
    entity_type: str
    entity_id: str
    revision: int


class PullResponse(CarnetSchema):
    records: list[RecordEnvelope]
    next_cursor: int
    has_more: bool


class PushResponse(CarnetSchema):
    applied: list[AppliedRecord]
    # Envelopes the server kept because it holds a newer version. The client
    # applies them, overwriting what it tried to push.
    conflicts: list[RecordEnvelope]
    cursor: int


class SyncStatusResponse(CarnetSchema):
    server_revision: int
    record_count: int
